### PR TITLE
Allure Report tab in build results.

### PR DESCRIPTION
**Pull request by smecsia**
# Summary

- Collect artifacts from all build stages and build report based on all artifacts of the build chain.
- Ability to download Allure executable if not exists.
- Different configuration options.
- Support for Amazon S3 and Server-local artifact handlers.

# Screenshots

### CODE REVIEW

**Comment by eroshenkoam** [comment]
👍

File: pom.xml [hunk 1]
```diff
@@ -1,69 +1,83 @@
...
             <groupId>junit</groupId>
             <artifactId>junit</artifactId>
-            <version>4.10</version>
+            <version>4.12</version>
             <scope>test</scope>
         </dependency>
-
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-all</artifactId>
```
Comment (baev) [inline pom.xml]: use `mockito-core` instead

File: pom.xml [hunk 2]
```diff
@@ -1,69 +1,83 @@
...
+            <version>1.10.19</version>
+        </dependency>
+        <dependency>
+            <groupId>org.hamcrest</groupId>
+            <artifactId>hamcrest-all</artifactId>
```
Comment (baev) [inline pom.xml]: use `hamcrest-core` instead

File: src/main/java/io/qameta/allure/bamboo/AllureConstants.java [hunk 3]
```diff
@@ -3,23 +3,45 @@
 /**
  * The internal class with some constants needed to Allure plugin.
  */
-final class AllureConstants {
+public interface AllureConstants {
```
Comment (baev) [inline src/main/java/io/qameta/allure/bamboo/AllureConstants.java]: That is kind of bad practice to use constant interface

File: src/main/java/io/qameta/allure/bamboo/AllureDownloader.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            return Optional.of(downloadToFile);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return Optional.empty();
+    }
+
+    private URL buildAllureDownloadUrl(String version) throws MalformedURLException {
+        return new URL(
+                "https://dl.bintray.com/qameta/generic/io/qameta/allure/allure/" + version + "/allure-" + version + ".zip"
```
Comment (baev) [inline src/main/java/io/qameta/allure/bamboo/AllureDownloader.java]: can we use sysprop or something for url?

File: src/main/java/io/qameta/allure/bamboo/AllureExecutableProvider.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    public AllureExecutableProvider(BambooExecutablesManager bambooExecutablesManager, AllureDownloader allureDownloader) {
+        this.bambooExecutablesManager = bambooExecutablesManager;
+        this.allureDownloader = allureDownloader;
+    }
+
+    Optional<AllureExecutable> provide(AllureGlobalConfig globalConfig, String executableName) {
+        return bambooExecutablesManager.getExecutableByName(executableName)
+                .map(allureHomeDir -> {
+                    final Path cmdPath = Paths.get(allureHomeDir, "bin", "allure");
```
Comment (baev) [inline src/main/java/io/qameta/allure/bamboo/AllureExecutableProvider.java]: I guess we should run allure.bat on Windows?

File: src/main/java/io/qameta/allure/bamboo/util/ExceptionUtil.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package io.qameta.allure.bamboo.util;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+
+public enum ExceptionUtil {
```
Comment (baev) [inline src/main/java/io/qameta/allure/bamboo/util/ExceptionUtil.java]: Consider using final class with private constructor instead

File: src/main/java/io/qameta/allure/bamboo/AllureViewReportCondition.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        final String buildNumberString = (String) context.get("buildNumber");
+        if (buildKey != null && buildNumberString != null) {
+            try {
+                int buildNumber = Integer.parseInt(buildNumberString);
+                final ResultsSummary resultsSummary = this.resultsSummaryManager.getResultsSummary(PlanKeys.getPlanResultKey(buildKey, buildNumber));
+                if (resultsSummary != null) {
+                    final AllureBuildResult buildResult = AllureBuildResult.fromCustomData(resultsSummary.getCustomBuildData());
+                    return (buildResult.hasInfo() && (resultsSummary.isFinished() || resultsSummary.isNotBuilt()));
+                }
+            } catch (Exception ignored) {
```
Comment (baev) [inline src/main/java/io/qameta/allure/bamboo/AllureViewReportCondition.java]: Log exception?

File: src/main/java/io/qameta/allure/bamboo/AllureReportServlet.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            final String filePath = matcher.group(3);
+            final PlanResultKey planResultKey = getPlanResultKey(planKeyString, parseInt(buildNumber));
+            ofNullable(resultsSummaryManager.getResultsSummary(planResultKey)).ifPresent(results -> {
+                final AllureBuildResult uploadResult = fromCustomData(results.getCustomBuildData());
+                if (uploadResult.isSuccess()) {
+                    artifactsManager.getArtifactUrl(planKeyString, buildNumber, filePath).ifPresent(file -> {
+                        try (final InputStream input = new URL(file.getLeft()).openStream()) {
+                            final Path path = Paths.get(file.getLeft());
+                            response.setHeader("Content-Type", getServletContext().getMimeType(filePath));
+//                            response.setHeader("Content-Length", String.valueOf(file.getRight()));
```
Comment (baev) [inline src/main/java/io/qameta/allure/bamboo/AllureReportServlet.java]: remove comment

File: src/main/java/io/qameta/allure/bamboo/AllureConstants.java (hunk 3 above)
Comment (smecsia) [inline src/main/java/io/qameta/allure/bamboo/AllureConstants.java]: Bad practice is to expose constants in general, much better to have them private or at least to have getters for them. But unlike Bloch, I hate verbosity, which is meant in having final class with private constructor and public static modifiers for every constant.
If we don't expose the interface (let's make it package private), it's not an anti-pattern.

File: src/main/java/io/qameta/allure/bamboo/AllureViewReportCondition.java (hunk 7 above)
Comment (smecsia) [inline src/main/java/io/qameta/allure/bamboo/AllureViewReportCondition.java]: copy-paste from Bamboo sources :)

File: pom.xml [hunk 9]
```diff
@@ -70,13 +70,15 @@
         </dependency>
         <dependency>
             <groupId>org.mockito</groupId>
-            <artifactId>mockito-all</artifactId>
+            <artifactId>mockito-core</artifactId>
             <version>1.10.19</version>
```
Comment (baev) [inline pom.xml]: 2.7.22

**Comment by smecsia** [comment]
Sorry guys, it appeared that we have to sign the CLA (contributor license agreement) with Atlassian first. I'll handle all the bureaucracy stuff and will reopen this PR later.

**Comment by smecsia** [comment]
After only 2 weeks I've got an answer that Apache 2.0 license does not require CLA at all :man_facepalming:

**Comment by smecsia** [comment]
It would be great if you could publish this plugin as soon as possible!

**Pull request merged**